persistence-cassandra-3.11/src/main/java/org/apache/cassandra/metrics/CassandraMetricsRegistry.java

```diff
@@ -0,0 +1,420 @@
+package org.apache.cassandra.metrics;
+
+import java.lang.reflect.Method;
+
+import com.codahale.metrics.Counter;
+import com.codahale.metrics.Histogram;
+import com.codahale.metrics.Meter;
+import com.codahale.metrics.Metric;
+import com.codahale.metrics.MetricRegistry;
+import com.codahale.metrics.Timer;
+import javax.management.MalformedObjectNameException;
+import javax.management.ObjectName;
+
+/**
+ * Patches the Cassandra metric implementation to allow injecting our own registry.
+ *
+ * <p>This is messy because of how the original class was designed (no interface contract, relies on
+ * static state...): {@link #actualRegistry} must be set manually before hitting any Cassandra code
+ * that might attempt to register a metric.
+ *
+ * <p>This class must mimic the exact public API of the original one.
+ */
+public class CassandraMetricsRegistry extends MetricRegistry {
```

Comment:
This is really ugly, but it's the simplest way to intercept metrics without too many impacts on the Cassandra code itself. I hope redefining `org.apache.cassandra` types in our own codebase is OK from a licensing perspective.

The original class also creates a MBean for each metric out of the box. I removed that because I don't think we want JMX by default, but that could be readded if necessary.

Reply:
Is there a way to do this generically or will we need to reimplement for each persistence module?

Reply:
The class's public contract changes across Cassandra versions, so we'll have to redefine it every time.